File: src/test_runner/expect/toMatchInlineSnapshot.zig
pub fn toMatchInlineSnapshot(this: *Expect, globalThis: *JSGlobalObject, callFrame: *CallFrame) bun.JSError!JSValue {
    defer this.postMatch(globalThis);
    const thisValue = callFrame.this();
    const _arguments = callFrame.arguments_old(2);
    const arguments: []const JSValue = _arguments.ptr[0.._arguments.len];

    this.incrementExpectCallCounter();

    const not = this.flags.not;
    if (not) {
        const signature = comptime getSignature("toMatchInlineSnapshot", "", true);
        return this.throw(globalThis, signature, "\n\n<b>Matcher error<r>: Snapshot matchers cannot be used with <b>not<r>\n", .{});
    }

    var has_expected = false;
    var expected_string: ZigString = ZigString.Empty;
    var property_matchers: ?JSValue = null;
    switch (arguments.len) {
        0 => {},
        1 => {
            if (arguments[0].isString()) {
                has_expected = true;
                try arguments[0].toZigString(&expected_string, globalThis);
            } else if (arguments[0].isObject()) {
                property_matchers = arguments[0];
            } else {
                return this.throw(globalThis, "", "\n\nMatcher error: Expected first argument to be a string or object\n", .{});
            }
        },
        else => {
            if (!arguments[0].isObject()) {
                const signature = comptime getSignature("toMatchInlineSnapshot", "<green>properties<r><d>, <r>hint", false);
                return this.throw(globalThis, signature, "\n\nMatcher error: Expected <green>properties<r> must be an object\n", .{});
            }

            property_matchers = arguments[0];

            if (arguments[1].isString()) {
                has_expected = true;
                try arguments[1].toZigString(&expected_string, globalThis);
            }
        },
    }

    var expected = expected_string.toSlice(default_allocator);
    defer expected.deinit();

    const expected_slice: ?[]const u8 = if (has_expected) expected.slice() else null;

    const value = try this.getValue(globalThis, thisValue, "toMatchInlineSnapshot", "<green>properties<r><d>, <r>hint");
    return this.inlineSnapshot(globalThis, callFrame, value, property_matchers, expected_slice, "toMatchInlineSnapshot");
}

const bun = @import("bun");
const ZigString = bun.ZigString;
const default_allocator = bun.default_allocator;

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/test_runner/expect/toMatchInlineSnapshot.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};
use bun_str::ZigString;

use super::Expect;

#[bun_jsc::host_fn(method)]
pub fn to_match_inline_snapshot(
    this: &mut Expect,
    global: &JSGlobalObject,
    frame: &CallFrame,
) -> JsResult<JSValue> {
    // TODO(port): `defer this.postMatch(globalThis)` — scopeguard here overlaps with &mut this;
    // Phase B should reshape (raw-ptr guard or RAII helper on Expect).
    scopeguard::defer! { this.post_match(global); }

    let this_value = frame.this_value();
    let arguments: &[JSValue] = frame.arguments_old(2);

    this.increment_expect_call_counter();

    let not = this.flags.not;
    if not {
        let signature = Expect::get_signature("toMatchInlineSnapshot", "", true);
        return this.throw(
            global,
            signature,
            format_args!(
                "\n\n<b>Matcher error<r>: Snapshot matchers cannot be used with <b>not<r>\n"
            ),
        );
    }

    let mut has_expected = false;
    let mut expected_string: ZigString = ZigString::EMPTY;
    let mut property_matchers: Option<JSValue> = None;
    match arguments.len() {
        0 => {}
        1 => {
            if arguments[0].is_string() {
                has_expected = true;
                arguments[0].to_zig_string(&mut expected_string, global)?;
            } else if arguments[0].is_object() {
                property_matchers = Some(arguments[0]);
            } else {
                return this.throw(
                    global,
                    "",
                    format_args!(
                        "\n\nMatcher error: Expected first argument to be a string or object\n"
                    ),
                );
            }
        }
        _ => {
            if !arguments[0].is_object() {
                let signature = Expect::get_signature(
                    "toMatchInlineSnapshot",
                    "<green>properties<r><d>, <r>hint",
                    false,
                );
                return this.throw(
                    global,
                    signature,
                    format_args!(
                        "\n\nMatcher error: Expected <green>properties<r> must be an object\n"
                    ),
                );
            }

            property_matchers = Some(arguments[0]);

            if arguments[1].is_string() {
                has_expected = true;
                arguments[1].to_zig_string(&mut expected_string, global)?;
            }
        }
    }

    let expected = expected_string.to_slice();
    // `defer expected.deinit()` — handled by Drop on the returned slice guard.

    let expected_slice: Option<&[u8]> = if has_expected { Some(expected.slice()) } else { None };

    let value = this.get_value(
        global,
        this_value,
        "toMatchInlineSnapshot",
        "<green>properties<r><d>, <r>hint",
    )?;
    this.inline_snapshot(
        global,
        frame,
        value,
        property_matchers,
        expected_slice,
        "toMatchInlineSnapshot",
    )
}

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     src/test_runner/expect/toMatchInlineSnapshot.zig (64 lines)
//   confidence: medium
//   todos:      1
//   notes:      defer post_match needs borrowck reshape; get_signature assumed const fn -> &'static str
// ──────────────────────────────────────────────────────────────────────────
